### PR TITLE
update weld to 3.1.7 and 4.0.1

### PR DESCRIPTION
**Pull request by benjamin-confino**

### CODE REVIEW

**Comment by benjamin-confino** [comment]
fixes #13552

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

File: dev/com.ibm.ws.cdi.1.2.weld/bnd.bnd [hunk 1]
```diff
@@ -19,7 +19,9 @@ Bundle-Description:CDI 1.2 Weld Integration, version ${bVersion}
 -dsannotations: com.ibm.ws.cdi.impl.DefaultSecurityService
 
 Export-Package: com.ibm.ws.cdi.executor, \
-                com.ibm.ws.cdi.utils
+                com.ibm.ws.cdi.utils, \
+                com.ibm.ws.cdi.impl.weld, \
```
Comment (tevans78) [inline dev/com.ibm.ws.cdi.1.2.weld/bnd.bnd]: i don't think you want to export this package from here. Put the proxy class in it's own new package and export that from here and the 2.0 version

File: dev/com.ibm.ws.cdi.1.2.weld/src/com/ibm/ws/cdi/impl/weld/package-info.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ * IBM Corporation - initial API and implementation
+ *******************************************************************************/
+
+@org.osgi.annotation.versioning.Version("1.0")
+package com.ibm.ws.cdi.impl.weld;
```
Comment (tevans78) [inline dev/com.ibm.ws.cdi.1.2.weld/src/com/ibm/ws/cdi/impl/weld/package-info.java]: don't use this same package name, use a new one

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

File: dev/com.ibm.ws.cdi.1.2.weld/src/com/ibm/ws/cdi/proxy/ProxyServicesImpl.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+/**
+ * This service is used to load proxy classes. We need a special classloader so that
+ * we can load both weld classes and app classes.
+ *
+ *
+ */
+public class ProxyServicesImpl extends AbstractProxyServices implements ProxyServices {
+
+    public ProxyServicesImpl() {
+        System.out.println("GREP 1.2");
```
Comment (tevans78) [inline dev/com.ibm.ws.cdi.1.2.weld/src/com/ibm/ws/cdi/proxy/ProxyServicesImpl.java]: don't forget to remove this

File: dev/io.openliberty.org.jboss.weld4/bnd.overrides [hunk 4]
```diff
@@ -8,16 +8,16 @@ Build-Identifier: SNAPSHOT-${now}
...
...;version="4.0.1";uses:="jakarta.el,jakarta.enterprise.context.spi,jakarta.enterprise.event,jakarta.enterprise.inject,jakarta.enterprise.inject.spi,jakarta.enterprise.inject.spi.configurator,org.jboss.weld.bean,org.jboss.weld.bean.proxy,org.jboss.weld.bootstrap.api,org.jboss.weld.bootstrap.spi,org.jboss.weld.construction.api,org.jboss.weld.ejb.spi,org.jboss.weld.event,org.jboss.weld.events,org.jboss.weld.inject,org.jboss.weld.interceptor.spi.model,org.jboss.weld.manager.api,org.jboss.weld.serialization.spi",org.jboss.weld.security;version="4.0.1";uses:="org.jboss.weld.security.spi",org.jboss.weld.util;version="4.0.1";uses:="jakarta.el,jakarta.enterprise.context.spi,jakarta.enterprise.event,jakarta.enterprise.inject,jakarta.enterprise.inject.spi,jakarta.enterprise.util,jakarta.interceptor,org.jboss.weld.bean,org.jboss.weld.bootstrap,org.jboss.weld.bootstrap.api,org.jboss.weld.bootstrap.spi,org.jboss.weld.exceptions,org.jboss.weld.inject,org.jboss.weld.injection,org.jboss.weld.interceptor.spi.model,org.jboss.weld.manager,org.jboss.weld.resources.spi,org.jboss.weld.serialization.spi",org.jboss.weld.xml;version="4.0.1";uses:="org.jboss.weld.bootstrap.spi,org.xml.sax,org.xml.sax.helpers"
```
Comment (tevans78) [inline dev/io.openliberty.org.jboss.weld4/bnd.overrides:20]: was this section kept the same as before or re-copied from the 4.0.1 weld osgi bundle?

File: dev/com.ibm.ws.cdi.2.0.weld/src/com/ibm/ws/cdi/impl/weld/AbstractProxyServices.java [hunk 5]
```diff
@@ -117,6 +120,8 @@ public ClassLoader run() {
                 if (cl instanceof BundleReference) {
                     Bundle b = ((BundleReference) cl).getBundle();
                     addWeldDynamicImports(b, WELD_PACKAGES);
+                } else if (cl == null) {
+                    return CLASS_LOADER_FOR_SYSTEM_CLASSES;
```
Comment (tevans78) [inline dev/com.ibm.ws.cdi.2.0.weld/src/com/ibm/ws/cdi/impl/weld/AbstractProxyServices.java]: put some comments here... and actually I don't see a reason why this code wouldn't be kept the same for the older version too?

File: dev/com.ibm.ws.cdi.2.0.weld/src/com/ibm/ws/cdi/impl/weld/AbstractProxyServices.java [hunk 6]
```diff
@@ -155,6 +160,19 @@ public ClassLoader run() {
...
                 return clazz;
+
+             } catch (LinkageError e) {
+                int tries = 10;
+                while (tries > 0) {
+                    Class<?> clazz = loadClass​(className, loader);
+                    if (clazz != null) {
+                        return clazz;
+                    }
+                    Thread.sleep(1000);
```
Comment (tevans78) [inline dev/com.ibm.ws.cdi.2.0.weld/src/com/ibm/ws/cdi/impl/weld/AbstractProxyServices.java]: definitely put comments here ... I don't like a Thread.sleep really

**Comment by benjamin-confino** [comment]
#build

File: dev/com.ibm.ws.cdi.2.0.weld/src/com/ibm/ws/cdi/impl/weld/AbstractProxyServices.java [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+        }
+
+        synchronized (classLoaderLock) {
+            try {
+                //First check we haven't defined this in another thread.
+                return loadClass​(className, loader);
+            } catch (ClassNotFoundException e) {
+                //Do nothing, move on to defining the class. 
+            }
+try {
```
Comment (tevans78) [inline dev/com.ibm.ws.cdi.2.0.weld/src/com/ibm/ws/cdi/impl/weld/AbstractProxyServices.java]: indentation!

File: dev/com.ibm.ws.cdi.2.0.weld/src/com/ibm/ws/cdi/impl/weld/AbstractProxyServices.java (hunk 6 above)
Comment (Emily-Jiang) [reply dev/com.ibm.ws.cdi.2.0.weld/src/com/ibm/ws/cdi/impl/weld/AbstractProxyServices.java]: Why does it need the sleep anyway?

File: dev/com.ibm.ws.cdi.1.2.weld/src/com/ibm/ws/cdi/impl/weld/package-info.java (hunk 2 above)
Comment (benjamin-confino) [reply dev/com.ibm.ws.cdi.1.2.weld/src/com/ibm/ws/cdi/impl/weld/package-info.java]: This is adding a missing package info file for the weld package.

**Comment by benjamin-confino** [comment]
#build

File: dev/io.openliberty.org.jboss.weld4/bnd.overrides (hunk 4 above)
Comment (benjamin-confino) [reply dev/io.openliberty.org.jboss.weld4/bnd.overrides:20]: I did a find/replace on the existing one. But the two manifests the osgi bnudle jar look like they are identical save version numbers

**Comment by benjamin-confino** [comment]
#build

File: dev/com.ibm.ws.cdi.2.0.weld/src/com/ibm/ws/cdi/impl/weld/AbstractProxyServices.java [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+	}
+
+	/*    @FFDCIgnore(ClassNotFoundException.class)
+    public Class<?> loadClass​(Class<?> originalClass, String classBinaryName) throws ClassNotFoundException {
+         if (originalClassToProxyClass.containsKey(originalClass)) {
+             return originalClassToProxyClass.get(originalClass);
+         } else {
+             throw new ClassNotFoundException("We failed to find a proxy class " + classBinaryName);
+         }
+    }*/
```
Comment (Azquelt) [inline dev/com.ibm.ws.cdi.2.0.weld/src/com/ibm/ws/cdi/impl/weld/AbstractProxyServices.java]: remove commented out code.

File: dev/com.ibm.ws.org.jboss.weld3/bnd.overrides [hunk 9]
```diff
@@ -3,18 +3,22 @@ bVersion=1.0
...
 Specification-Version: 2.0
 DynamicImport-Package: *
 Export-Package: !org.jboss.weld.environment.se.*, \
- org.jboss.*;version="3.1.4.Final", \
- org.jboss.weldx.transaction;version="1.0.16"
-Private-Package: org.jboss.weld.environment.se.*
+ org.jboss.*;version="3.1.7.SP1", \
+ org.jboss.weldx.transaction;version="1.0.16", \
+ !org.jboss.weld.bean.proxy.util\
+ !org.jboss.weld.proxy
```
Comment (Azquelt) [inline dev/com.ibm.ws.org.jboss.weld3/bnd.overrides:18]: Why this change. Were these packages unexported in the 3.1.7.SP1 bundle manifest?

**Review by Azquelt — APPROVED** [review]

File: dev/com.ibm.ws.org.jboss.weld3/bnd.overrides (hunk 9 above)
Comment (benjamin-confino) [reply dev/com.ibm.ws.org.jboss.weld3/bnd.overrides:18]: I think that was just an attempt to cut the number of exports.

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#run-libby-bot
